tests/tracing.rs

```diff
@@ -0,0 +1,147 @@
+//! SOFTWARE.
+
+use std::io;
+use std::sync::{Arc, Mutex, MutexGuard};
+
+use tracing::{Dispatch, Instrument};
+use tracing_subscriber::fmt::MakeWriter;
+use tracing_subscriber::FmtSubscriber;
+use xtra::prelude::*;
+
+#[derive(Debug)]
+pub struct MockWriter {
+    buf: Arc<Mutex<Vec<u8>>>,
+}
+
+impl MockWriter {
+    /// Create a new `MockWriter` that writes into the specified buffer (behind a mutex).
+    pub fn new(buf: Arc<Mutex<Vec<u8>>>) -> Self {
+        Self { buf }
+    }
+
+    /// Give access to the internal buffer (behind a `MutexGuard`).
+    fn buf(&self) -> io::Result<MutexGuard<Vec<u8>>> {
+        // Note: The `lock` will block. This would be a problem in production code,
+        // but is fine in tests.
+        self.buf
+            .lock()
+            .map_err(|_| io::Error::from(io::ErrorKind::Other))
+    }
+}
+
+impl io::Write for MockWriter {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        // Lock target buffer
+        let mut target = self.buf()?;
+
+        // Write to stdout in order to show up in tests
+        print!("{}", String::from_utf8(buf.to_vec()).unwrap());
+
+        // Write to buffer
+        target.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.buf()?.flush()
+    }
+}
+
+impl MakeWriter<'_> for MockWriter {
+    type Writer = Self;
+
+    fn make_writer(&self) -> Self::Writer {
+        MockWriter::new(self.buf.clone())
+    }
+}
+
+/// Return a new subscriber that writes to the specified [`MockWriter`].
+///
+/// [`MockWriter`]: struct.MockWriter.html
+pub fn get_subscriber(mock_writer: MockWriter, env_filter: &str) -> Dispatch {
+    FmtSubscriber::builder()
+        .with_env_filter(env_filter)
+        .with_writer(mock_writer)
+        .with_level(true)
+        .with_ansi(false)
+        .without_time()
+        .into()
+}
+
+pub fn with_logs<F>(buf: &Mutex<Vec<u8>>, f: F)
+where
+    F: Fn(&[&str]),
+{
+    let buf = buf.lock().unwrap();
+    let logs: Vec<&str> = std::str::from_utf8(&buf)
+        .expect("Logs contain invalid UTF8")
+        .lines()
+        .collect();
+    f(&logs)
+}
+
+struct Tracer;
+
+#[async_trait]
+impl Actor for Tracer {
+    type Stop = ();
+    async fn stopped(self) {}
+}
+
+struct Hello(&'static str);
+
+#[async_trait]
+impl Handler<Hello> for Tracer {
+    type Return = ();
+
+    async fn handle(&mut self, message: Hello, _ctx: &mut Context<Self>) {
+        tracing::info!("Hi {}", message.0)
+    }
+}
+
+#[tokio::test]
+async fn assert_send_is_child_of_span() {
+    let buf = Arc::new(Mutex::new(vec![]));
+    let mock_writer = MockWriter::new(buf.clone());
+    let subscriber = get_subscriber(mock_writer, "tracing=trace,xtra=trace");
+    let _g = tracing::dispatcher::set_default(&subscriber);
+
+    let (addr, mut ctx) = Context::<Tracer>::new(None);
+    let _ = addr
+        .send(Hello("hi"))
+        .split_receiver()
```

Comment:
I find the usage of `split_receiver` a bit noisy here. Why not just use `spawn_global` here?

Using `split_receiver` here seems like it matters for the test.

Reply:
> I find the usage of split_receiver a bit noisy here. Why not just use spawn_global here?

The default subscriber is thread local, I believe, so we need to handle the message on the same thread. It might also change with tokio task - I'm not sure. To make sure everything definitely worked, I wanted to do everything on the same task and thread.

Reply:
> > I find the usage of split_receiver a bit noisy here. Why not just use spawn_global here?
> 
> The default subscriber is thread local, I believe, so we need to handle the message on the same thread. It might also change with tokio task - I'm not sure. To make sure everything definitely worked, I wanted to do everything on the same task and thread.

I believe the default tokio test executor is single-threaded and if not, you can customise it via the `tokio::test` attribute.

Reply:
Yep, it is single-threaded: https://docs.rs/tokio/latest/tokio/attr.test.html#using-default

Reply:
I've changed it now